Ensure signals are handled before Unix.kill returns

stedolan opened the pull request:
POSIX does not make very many guarantees about when signals are handled, but one guarantee that it does make is that if a process sends _itself_ a signal (using e.g. `kill(getpid(), ...)`), then signals will be handled before `kill` returns.

This is currently not true of OCaml's `Unix.kill`, this patch makes it so.

(This came up during testing of #9722, but is a separate issue)

stedolan committed:
Ensure signals are handled before Unix.{kill,sigprocmask} return

jhjourdan reviewed (approved):
The change is trivial, looks good to me.

stedolan commented:
Searching through the standard, I see that POSIX makes the same stipulation about the other kills that can cause a signal to become pending (sigprocmask, pthread_sigmask), so I'll fix those too.

stedolan commented:
@jhjourdan it turns out that your test from #9027 (tests/callback/signals_alloc.ml) is explicitly testing that `kill` does _not_ run the signal handler. Is this intentional?

stedolan commented:
I've updated the tests to avoid assuming this behaviour of kill, and added a test for sigprocmask. (This test passes with no code change due to a poll in caml_leave_blocking_section, but I want the test there before #9722  is merged)

jhjourdan commented:
> @jhjourdan it turns out that your test from #9027 (tests/callback/signals_alloc.ml) is explicitly testing that kill does not run the signal handler. Is this intentional?

Yes, it is. The point of this test is to check that allocations do indeed poll, so polling should not occur in when raising the signal. Perhaps you could use the same strategy as in the other test?

stedolan commented:
> Yes, it is. The point of this test is to check that allocations do indeed poll, so polling should not occur in when raising the signal. Perhaps you could use the same strategy as in the other test?

Done.

nojb commented:
This PR has broken the INRIA CI:

- https://ci.inria.fr/ocaml/job/Risc-V/244/
- https://ci.inria.fr/ocaml/job/main/flambda=false,label=ocaml-ppc-64-be/2057/
- https://ci.inria.fr/ocaml/job/main/flambda=false,label=ocaml-arm-64/2057/

The error (visible in the above logs) is
> Fatal error: exception File "signals_alloc.ml", line 27, characters 2-8: Assertion failed

stedolan commented:
That is weird. I'm running a precheck build with a more verbose version of the test to see if I can see what's going on.

stedolan commented:
Possible fix in #9814
